### PR TITLE
fix: Add debounce for 300 ms when searching

### LINKED ISSUE #1027: Searching bogs down while typing a search term
Reproducing:

1) Use redoc on a large spec.

2) Quickly type a phrase into the search bar

Result:

The UI hangs, a search is performed for each change event.

Fix:

Debounce searches, so a search is only performed when a certain amount of time has passed since the last change event.

### PR DESCRIPTION
**Pull request by faheem00**
Fixes #1027 

Steps taken
- Ran tslint
- Ran all tests
- Ran `yarn prettier`

### CODE REVIEW

**Comment by faheem00** [comment]
`Search should support arrow navigation` test seems to be failing. Is it because the search result function was debounced, and the result display is delayed?

File: src/components/SearchBox/SearchBox.tsx [hunk 1]
```diff
@@ -106,13 +107,25 @@ export class SearchBox extends React.PureComponent<SearchBoxProps, SearchBoxStat
...
+          }),
+        300,
+      ),
+    );
+  };
 
-    this.props.search.search(event.target.value).then(res => {
-      this.setResults(res, q);
-    });
+  debounce = (fn: (..._) => any, wait: number = 1): (() => void) => {
```
Comment (RomanHotsiy) [inline src/components/SearchBox/SearchBox.tsx]: Could you remove your hand-made debounce and use `debounce` from `decko` (we already use this dependency) instead?

Comment (faheem00) [reply src/components/SearchBox/SearchBox.tsx]: I tried using the `debounce` implementation from decko as such - 
```javascript
@debounce
search = (event: React.ChangeEvent<HTMLInputElement>) => {
...
}
```
But this showed the following warning when trying to compile- 
`TS1240: The return type of a property decorator function must be either 'void' or 'any'. `

Furthermore, when I try to use `delay` as argument as such - 
```javascript
@debounce(300)
search = (event: React.ChangeEvent<HTMLInputElement>) => {
...
}
```
I get the following warning - 
`TS1240: 
Unable to resolve signature of property decorator when called as an expression.`

As the script was not compiling, I opted to use a hand-made debounce. One thing to note is that the files in which the `debounce` from `decko` is already being used (e.g. `ScrollService.ts` and `HistoryService.ts` are `ts` file, whereas `SearchBox` is a `tsx` file. Maybe there needs to be a change in lint configuration for `tsx` files?

Comment (RomanHotsiy) [reply src/components/SearchBox/SearchBox.tsx]: Use 

```
  @bind
  @debounce
  search(event: React.ChangeEvent<HTMLInputElement>) {
```

Note `search` is not field with lambda function as value but method.

Comment (faheem00) [reply src/components/SearchBox/SearchBox.tsx]: Pushed the changes

**Comment by fabienpomerol** [comment]
@faheem00  guys any update on this subject ? i'm also interested by the fix :)

**Pull request merged**